[4.x] Schema queries with configurable server-side timeouts

Bouncheck opened the pull request:
Adds USING TIMEOUT clause to schema queries when applicable.
Uses the same timeout as driver-side timeouts defined by `METADATA_SCHEMA_REQUEST_TIMEOUT` ("advanced.metadata.schema.request-timeout").
Fixes #298

mykaul commented on core/src/main/java/com/datastax/oss/driver/api/core/config/DefaultDriverOption.java:
```diff
@@ -612,12 +612,20 @@ public enum DefaultDriverOption implements DriverOption {
    */
   METADATA_SCHEMA_REQUEST_TIMEOUT("advanced.metadata.schema.request-timeout"),
+  /**
+   * The Scylla side timeout (in seconds) for the requests to the schema tables. Applied by adding
```
```suggestion
   * The server side timeout (in seconds) for the requests to the schema tables. Applied by adding
```

mykaul commented on core/src/main/java/com/datastax/oss/driver/api/core/config/DefaultDriverOption.java:
```diff
@@ -612,12 +612,20 @@ public enum DefaultDriverOption implements DriverOption {
   METADATA_SCHEMA_REQUEST_TIMEOUT("advanced.metadata.schema.request-timeout"),
+  /**
+   * The Scylla side timeout (in seconds) for the requests to the schema tables. Applied by adding
+   * USING TIMEOUT to schema queries. Works with Scylla versions 5.0.0/2022.1.0 or newer.
```
```suggestion
   * USING TIMEOUT to schema queries. Works with ScyllaDB versions 5.0.0/2022.1.0 or newer.
```

mykaul commented on core/src/main/java/com/datastax/oss/driver/api/core/config/DefaultDriverOption.java:
```diff
@@ -612,12 +612,20 @@ public enum DefaultDriverOption implements DriverOption {
+   *
+   * <p>Value-type: {@link java.time.Duration Duration}
+   */
+  METADATA_SCHEMA_REQUEST_SCYLLA_TIMEOUT("advanced.metadata.schema.request-scylla-timeout"),
```
It doesn't work against Cassandra? Is it a a ScyllaDB only parameter?

mykaul commented on core/src/main/java/com/datastax/oss/driver/api/core/metadata/Node.java:
```diff
@@ -137,6 +137,17 @@ public interface Node {
+  /**
+   * The Scylla version of the server.
+   *
+   * <p>Should be non-null on control node when connected to newer versions of Scylla. Driver will
```
newer? newer than?

mykaul commented on core/src/main/java/com/datastax/oss/driver/internal/core/metadata/DefaultTopologyMonitor.java:
```diff
@@ -205,7 +207,22 @@ public CompletionStage<Iterable<NodeInfo>> refreshNodeList() {
+                      })
+                  .toCompletableFuture()
+                  .join();
+          System.out.println("control con scylla version string: " + scyllaVersion);
```
debug leftovers?

Bouncheck replied:
Yes, missed that one

mykaul commented:
I don't fully understand the whole Scylla version issue here. 5.0/2022.1 are ancient version. If someone is using such an old version, what happens when we try to fetch the schema with timeout?

Bouncheck commented:
> I don't fully understand the whole Scylla version issue here. 5.0/2022.1 are ancient version. If someone is using such an old version, what happens when we try to fetch the schema with timeout?

I couldn't figure out a clean way to get Scylla version without querying `system.versions` which was introduced in those versions. In this change if we cannot determine Scylla version we fall back to relying on reported Cassandra version. If the Scylla version is too old (or undetermined) for using timeout we just don't use it.

I probably should've marked this as draft for now since I'm still cleaning this up. I may switch to either checking for shard awareness (as a proxy check for "is it scylla cluster") or running trial query with `using timeout` instead of checking exact Scylla version

mykaul commented:
> > I don't fully understand the whole Scylla version issue here. 5.0/2022.1 are ancient version. If someone is using such an old version, what happens when we try to fetch the schema with timeout?
> 
> I couldn't figure out a clean way to get Scylla version without querying `system.versions` which was introduced in those versions. In this change if we cannot determine Scylla version we fall back to relying on reported Cassandra version. If the Scylla version is too old (or undetermined) for using timeout we just don't use it.

What do you mean too old? What happen when you do try to use timeout, and the scylla version is too old? If the timeout is ignored - fine, go with it. Doesn't matter. If it crashes, I'm more worried. If it errs the CQL query - then I think you should fallback to not using timeout.

> 
> I probably should've marked this as draft for now since I'm still cleaning this up. I may switch to either checking for shard awareness (as a proxy check for "is it scylla cluster") or running trial query with `using timeout` instead of checking exact Scylla version

Why? What would happen when you use it against a Cassandra cluster? Same as the above, IMHO (just as we try initially with high DSE versions that are not supported against Scylla!)

Bouncheck commented:
Cassandra does not have `USING TIMEOUT`. Select queries do not allow `USING` altogether. Both will result in SyntaxException. I have to check how older versions of Scylla react, I've just picked the oldest supporting according to commits (4.4 was first with using timeout cql extension, 5.0 was for `system.versions` table containing scylla version).

mykaul commented:
> Cassandra does not have `USING TIMEOUT`. Select queries do not allow `USING` altogether. Both will result in SyntaxException. I have to check how older versions of Scylla react, I've just picked the oldest supporting according to commits (4.4 was first with using timeout cql extension, 5.0 was for `system.versions` table containing scylla version).

I'd catch the exception and re-run the query without 'USING TIMEOUT' . Probably a more optimistic approach (happy path) then the whole additional query we have. Of course, we may use the Scylla version or whatnot elsewhere, later, etc, so it might have value by itself.

Bouncheck committed:
Add USING TIMEOUT to CassandraSchemaQueries

The driver will now use existing configurable option
`METADATA_SCHEMA_REQUEST_TIMEOUT` in a `USING TIMEOUT` clause
with each schema query (if applicable).
This will make it a server side timeout in addition to driver side.
The driver will apply the `USING` clause if the sharding info of control
connection channel is not null. This is used as a proxy check to differentiate
between Scylla and Cassandra.
The behaviour should remain unchanged for Cassandra clusters.
USING TIMEOUT is a ScyllaDB CQL extension and it is available in
versions 4.4/2022.1.0 or newer.

Bouncheck commented:
I've switched to running small trial select with `USING TIMEOUT` clause. If it completes without exceptions then the following schema queries will use this clause with the same configurable timeout as it is for driver side timeout. I've also removed unnecessary new additions.

Commit with querying for Scylla version remains as it was. While it is no longer necessary for this change I also believe it may be useful later and its impact should be minimal.

Waiting to see CI results.

mykaul commented on core/src/main/java/com/datastax/oss/driver/api/core/metadata/Node.java:
```diff
@@ -15,6 +15,11 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
+/*
```
I don't understand this change - it's inconsistent - in some files it was added, in others it wasn't.

mykaul commented on core/src/main/java/com/datastax/oss/driver/internal/core/metadata/DefaultTopologyMonitor.java:
```diff
@@ -173,7 +173,9 @@ public CompletionStage<Iterable<NodeInfo>> refreshNodeList() {
-
+    CompletionStage<AdminResult> localVersionsQuery =
+        query(channel, "SELECT version FROM system.versions WHERE key = 'local'");
+    // localVersionsQuery.whenComplete()
```
?

mykaul commented on core/src/main/java/com/datastax/oss/driver/internal/core/metadata/NodeInfo.java:
```diff
@@ -106,6 +106,17 @@ public interface NodeInfo {
+  /**
+   * The Scylla version that this node runs.
+   *
+   * <p>This may be used for schema queries. There are Scylla-specific extensions like USING TIMEOUT
```
```suggestion
   * <p>This may be used for schema queries. There are Scylla-specific extensions such as USING TIMEOUT
```

mykaul commented on core/src/main/java/com/datastax/oss/driver/api/core/metadata/Node.java:
```diff
@@ -15,6 +15,11 @@
  * limitations under the License.
  */
+/*
+ * Copyright (C) 2023 ScyllaDB
```
2023 here, 2024 elsewhere?

Bouncheck replied:
Whenever I noticed it's missing i tried to apply the earliest year I could see in git blame

mykaul commented on core/src/test/java/com/datastax/oss/driver/internal/core/metadata/DefaultTopologyMonitorTest.java:
```diff
@@ -479,7 +486,12 @@ public void should_warn_when_control_host_found_in_system_peers_v2() {
     topologyMonitor.stubQueries(
         new StubbedQuery("SELECT * FROM system.local", mockResult(local)),
-        new StubbedQuery("SELECT * FROM system.peers_v2", mockResult(peer3, peer2, peer1)));
+        new StubbedQuery("SELECT * FROM system.peers_v2", mockResult(peer3, peer2, peer1)),
```
btw, if peers_v2 failed, it means we are using Scylla ;-)

Bouncheck commented:
I've removed commit with `getScyllaVersion()` change for now. There are more test adjustments needed for it, besides missed license headers, etc. and it's blocking the main purpose of this PR. I'll save that one for another time.

Bouncheck commented:
I don't understand what's happening with gh actions. [Unit tests fail due to cancellation](https://github.com/scylladb/java-driver/actions/runs/9747387563/job/26899979958?pr=312) but I have no idea what is triggering it.

Bouncheck commented:
Oh, it's just timeout. I guess it became too small at some point.
edit: or not, since everything seems to hang on a test related to the change

mykaul commented on core/src/main/java/com/datastax/oss/driver/internal/core/metadata/schema/queries/CassandraSchemaQueries.java:
```diff
@@ -73,6 +81,10 @@ protected CassandraSchemaQueries(
             DefaultDriverOption.METADATA_SCHEMA_REFRESHED_KEYSPACES, Collections.emptyList());
     assert refreshedKeyspaces != null; // per the default value
     this.keyspaceFilter = KeyspaceFilter.newInstance(logPrefix, refreshedKeyspaces);
+    this.usingTimeoutClause =
```
Q: shouldn't USING TIMEOUT be only applied if METADATA_SCHEMA_REQUEST_TIMEOUT has some value? And if the default will have 0, then we can keep both backward compat. and Cassandra compat. and yet satisfy the ask of this issue, no?

Bouncheck replied:
This value was originally used as a driver side timeout only and by default it's set to 2 seconds. If set to 0 I believe every query would instantly time out driver-side. 
If i went back to introducing a new config value for this and made it the way you say, then yes it would satisfy the ask of this issue, but I still would have to run a test query for Scylla and adjust tests accordingly. Unless we shift the responsibility to user side and say "use this only if you know your cluster supports this". Cassandra tests would require only adding mocks for the config in that case. Scylla tests would be simpler because there would be no test query to mock.

mykaul replied:
I think (and others should chime in here) that it should be a Scylla only config, that should not be set by default (and if it's 0, we simply do not use it - or perhaps - though I dislike it, you can use -1 as disable and 0 as infinite - a bit ugly if you ask me). We will ask customers to use this value IF they have a large schema and schema pulls fail and retried (a log about this could be useful as well!). Then we are good from all sides - Cassandra wouldn't use this by default and there is no change in the default behavior.

Bouncheck replied:
I think it would be weird to have different timeout driver and server side (is there a use case for that?), but there's a middle ground here with adding boolean option for enabling Using Timeout clause. If enabled driver will pass the same timeout to query strings.

mykaul replied:
The whole point of the feature here is that *schema* pulls will have their own, specific (longer...) timeout values.

Bouncheck replied:
We do have that distinction already on the driver side. I thought the problem was that timeout is not being overridden on Scylla side by adding `USING TIMEOUT` to queries which causes large schema pulls to time out on Scylla side.

For global limits we have `REQUEST_TIMEOUT` and for schema we have `METADATA_SCHEMA_REQUEST_TIMEOUT` which is already implemented.
I may have caused some confusion by originally introducing a new very similarly named config for schema timeouts intended for server-side but I've removed it later from this PR.

mykaul replied:
So I'm somewhat confused what we are doing in this PR. Perhaps a quick call is needed.

mykaul replied:
(in short, why not add USING TIMEOUT to METADATA_SCHEMA_REQUEST_TIMEOUT ? This I think it is the original ask)

Bouncheck replied:
I think this is what we're doing now: adding `USING TIMEOUT` string to schema query strings with the value of `METADATA_SCHEMA_REQUEST_TIMEOUT` seconds.
The problem with implementation was that `USING TIMEOUT` is an extension to CQL not present in every Scylla version and not present in Cassandra at all. Adding it when not supported will cause queries to fail.
Originally I've had some additional commits with changes that were supposed to help with determining if it's supported by server but I've removed most of that because that had its own problems.

mykaul replied:
OK, let's sort it out:
1. USING TIMEOUT is supported in every reasonable and supported Scylla version. Period. Let's not look into older versions.
2. With Cassandra, I still think catching the exception (which you should get right away) and retrying without USING TIMEOUT would be easiest.
2b. You should be able to distinguish between Scylla and Cassandra easily, and I assume the driver already has this distinction somewhere (no. of shards for example?) - and then use USING TIMEOUT only for Scylla (again, regardless of the version).

roydahan replied:
This is what we already agreed on.
I don't understand what's the confusion here.

Bouncheck commented:
Unit tests should pass now

Bouncheck commented:
I couldn't understand why my test query was causing integration tests to take too long, so instead I look at sharding info of the control connection channel. If it's non-null I am assuming I'm dealing with Scylla and will use the timeouts. I don't check the Scylla version.
I see that some integration tests failed already but this time it's the flakiness of other stuff.

mykaul commented:
This looks much better - please resolve all comments.

Lorak-mmk reviewed (changes requested):
Commit message says `USING TIMESTAMP` instead of `USING TIMEOUT`
Apart from that looks good to me

Bouncheck commented:
I've fixed the typo @Lorak-mmk

Lorak-mmk reviewed (approved):
LGTM. One more thing: shouldn't we do the same with topology queries?

mykaul commented:
> LGTM. One more thing: shouldn't we do the same with topology queries?

Is topology response large? I doubt it, but we can certainly test it with a (mockup) large setup.

Lorak-mmk commented:
> > LGTM. One more thing: shouldn't we do the same with topology queries?
> 
> Is topology response large? I doubt it, but we can certainly test it with a (mockup) large setup.

There are as many rows as nodes in the cluster. What is the biggest cluster that we know of?

mykaul commented:
1. What's the default of the timeout?
2. Have we added this to the documentation?

Bouncheck commented:
> 1. What's the default of the timeout?
> 
> 2. Have we added this to the documentation?

1. The default is the same value as `basic.request.timeout` is set to (which is a global timeout for every query) and the default of that is 2 seconds.
2. No.
The documentation remained the same as before, so there is a documentation for the option but the new functionality was not documented. The fact that when using Scylla this option will override server's global timeouts is probably worth mentioning though.

mykaul commented:
> > 1. What's the default of the timeout?
> > 2. Have we added this to the documentation?
> 
> 1. The default is the same value as `basic.request.timeout` is set to (which is a global timeout for every query) and the default of that is 2 seconds.

Makes little sense to me. It means by default, users might have an issue. It should be 20 seconds or so. It should be high. Without setting a good new default to it, what will happen is that user bump into the issue first, then fix this and set some value.

> 2. No.
>    The documentation remained the same as before, so there is a documentation for the option but the new functionality was not documented. The fact that when using Scylla this option will override server's global timeouts is probably worth mentioning though.

It should be documented.

Bouncheck commented:
> Makes little sense to me. It means by default, users might have an issue. It should be 20 seconds or so. It should be high. Without setting a good new default to it, what will happen is that user bump into the issue first, then fix this and set some value.

My impression was that 2 seconds for a single query is already pretty high. 
The original issue was occurring because the timeout set on Scylla side was too small. I haven't heard of cases where 2 seconds in driver config was not enough, but to be honest I don't have much knowledge how users set up the driver and if it's an issue.

Anyway, if schema queries take over 10 seconds then that probably would not go unnoticed, so I [made a PR](https://github.com/scylladb/java-driver/pull/404) addressing both points (default value to 20s and documentation).

dkropachev commented:
Timeouts on the driver-side should not be lower than on server side, otherwise it leads to retry storm when cluster is overloaded.
